core/network-libp2p/src/custom_proto.rs

```diff
@@ -41,67 +41,203 @@ pub struct RegisteredProtocol<T> {
+	/// Returns the protocol id.
+	#[inline]
+	pub fn protocol_id(&self) -> ProtocolId {
+		self.protocol_id
+	}
+
+	/// Returns the version of the protocol that was negotiated.
+	#[inline]
+	pub fn protocol_version(&self) -> u8 {
+		self.protocol_version
+	}
+
+	/// Starts a graceful shutdown process on this substream.
+	///
+	/// Note that "graceful" means that we sent a closing message. We don't wait for any
+	/// confirmation from the remote.
+	///
+	/// After calling this, the stream is guaranteed to finish soon-ish.
+	pub fn shutdown(&mut self) {
+		self.is_closing = true;
+		if let Some(task) = self.to_notify.take() {
+			task.notify();
+		}
+	}
+
+	/// Sends a message to the substream.
+	pub fn send_message(&mut self, Packet { id: packet_id, data }: Packet) {
+		if packet_id >= self.packet_count {
+			error!(target: "sub-libp2p", "Tried to send a packet with an invalid ID {}", packet_id);
+			return;
+		}
+
+		let mut message = Bytes::with_capacity(1 + data.len());
+		message.extend_from_slice(&[packet_id]);
+		message.extend_from_slice(&data);
+		self.send_queue.push_back(message);
+
+		// If the length of the queue goes over a certain arbitrary threshold, we print a warning.
+		// TODO: figure out a good threshold
+		if self.send_queue.len() >= 2048 {
+			warn!(target: "sub-libp2p", "Queue of packets to send over substream is pretty \
+				large: {}", self.send_queue.len());
+		}
+
+		if let Some(task) = self.to_notify.take() {
+			task.notify();
+		}
+	}
+
+	/// Turns raw data into a packet and checks whether it is valid.
+	fn data_to_packet(&self, mut data: BytesMut) -> Result<Packet, ()> {
+		// The `data` should be prefixed by the packet ID, therefore an empty packet is invalid.
+		if data.is_empty() {
+			debug!(target: "sub-libp2p", "ignoring incoming packet because it was empty");
+			return Err(());
+		}
+
+		let packet = {
+			let id = data[0];
+			let data = data.split_off(1);
+			Packet { id, data: data.freeze() }
+		};
+
+		if packet.id >= self.packet_count {
+			debug!(target: "sub-libp2p", "ignoring incoming packet because packet_id {} is \
+				too large", packet.id);
+			return Err(())
+		}
+
+		Ok(packet)
+	}
+}
+
+impl<TSubstream> Stream for RegisteredProtocolSubstream<TSubstream>
+where TSubstream: AsyncRead + AsyncWrite,
+{
+	type Item = Packet;
+	type Error = IoError;
+
+	fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
+		// If we are closing, close as soon as the Sink is closed.
+		if self.is_closing {
+			return Ok(self.inner.close()?.map(|()| None));
+		}
+
+		// Flushing the local queue.
+		while let Some(packet) = self.send_queue.pop_front() {
+			match self.inner.start_send(packet)? {
+				AsyncSink::NotReady(packet) => {
+					self.send_queue.push_front(packet);
+					break;
+				},
+				AsyncSink::Ready => self.requires_poll_complete = true,
+			}
+		}
+
+		// Flushing if necessary.
+		if self.requires_poll_complete {
+			match self.inner.poll_complete()? {
```

Comment:
Maybe just `if let Async::Ready(()) = self.inner.poll_complete()? { self.requires_poll_complete = false; }`.
Reduces the amount of the code a little bit :D